lib/Sema/CSApply.cpp

```diff
@@ -126,6 +128,16 @@ static ConcreteDeclRef generateDeclRefForSpecializedCXXFunctionTemplate(
   auto *newParamList =
       ParameterList::create(ctx, SourceLoc(), newParams, SourceLoc());
 
+  if (isa<ConstructorDecl>(oldDecl)) {
+    DeclName ctorName(ctx, DeclBaseName::createConstructor(), newParamList);
+    auto newCtorDecl = ConstructorDecl::createImported(
+        ctx, specialized, ctorName, oldDecl->getLoc(), /*failable=*/false,
+        /*failabilityLoc=*/SourceLoc(), /*throws=*/false,
+        /*throwsLoc=*/SourceLoc(), newParamList, /*genericParams=*/nullptr,
+        oldDecl->getDeclContext());
+    return ConcreteDeclRef(newCtorDecl);
+  }
+
   // Generate a name for the specialized function.
   std::string newNameStr;
   llvm::raw_string_ostream buffer(newNameStr);
```

Comment:
@varungandhi-apple because I added a special case for constructor decls, we don't need to use `getMangledName`. So, thanks for telling me to wait in #34389.